### PR TITLE
Fix fuzz test and enable fuzz testing CI

### PR DESCRIPTION
**Pull request by magicalne**
Migrate fuzz testing from https://github.com/godwokenrises/godwoken-polyjuice/pull/181 and enable fuzz testing CI.

### CODE REVIEW

File: gwos-evm/polyjuice-tests/fuzz/gw-syscall-simulator/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,709 @@
...
+    }
+    fn get_data(&self, data_hash: &H256) -> Option<Bytes> {
+        self.state.get_data(data_hash)
+    }
+}
+
+static HOST: Lazy<Mutex<GodwokenHost>> = Lazy::new(|| Mutex::new(GodwokenHost::new()));
+
+#[no_mangle]
+pub extern "C" fn ckb_exit(code: i8) -> i32 {
```
Comment (blckngm) [inline gwos-evm/polyjuice-tests/fuzz/gw-syscall-simulator/src/lib.rs]: Nit: strictly speaking these functions should all be `unsafe`.

**Review by blckngm — DISMISSED** [review]

File: gwos-evm/polyjuice-tests/fuzz/gw-syscall-simulator/src/lib.rs (hunk 1 above)
Comment (blckngm) [reply gwos-evm/polyjuice-tests/fuzz/gw-syscall-simulator/src/lib.rs]: I mean all public functions that involves raw pointers.

Comment (magicalne) [reply gwos-evm/polyjuice-tests/fuzz/gw-syscall-simulator/src/lib.rs]: So you suggest turning all those functions into `unsafe extern` functions? So people know it is unsafe to use.

BTW, those `fn`s are supposed to be used in C code only, where safety is not the concern, I guess. Especially for the fuzz testing, I hope they crash. 🤪

Comment (blckngm) [reply gwos-evm/polyjuice-tests/fuzz/gw-syscall-simulator/src/lib.rs]: Yeah, I understand that these are supposed to be called only by C code. So it's just a suggestion.

Comment (magicalne) [reply gwos-evm/polyjuice-tests/fuzz/gw-syscall-simulator/src/lib.rs]: Done

**Review by blckngm — DISMISSED** [review]

File: gwos-evm/polyjuice-tests/fuzz/gw-syscall-simulator/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,714 @@
...
+    hasher.finalize(&mut data_hash);
+    // insert data hash into SMT
+    let data_hash_key = build_data_hash_key(&data_hash);
+    let mut host = HOST.lock().unwrap();
+    host.state
+        .update_raw(data_hash_key, H256::one())
+        .expect("gw store data");
+    host.state.insert_data(data_hash.into(), data);
+
+    host.run_result.write_data_hashes.insert(data_hash_key);
```
Comment (jjyr) [inline gwos-evm/polyjuice-tests/fuzz/gw-syscall-simulator/src/lib.rs]: https://github.com/godwokenrises/godwoken/blob/develop/crates/generator/src/syscalls/mod.rs#L411

`data_hash_key` should not be stored into `write_data_hashes`

File: gwos-evm/polyjuice-tests/fuzz/gw-syscall-simulator/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,714 @@
...
+        Err(_err) => {
+            return ERROR;
+        }
+    };
+
+    let account_nonce_key = build_account_field_key(id, GW_ACCOUNT_NONCE_TYPE);
+    host.state
+        .update_raw(account_nonce_key, H256::zero())
+        .expect("account nonce key");
+    host.run_result.write_data_hashes.insert(account_nonce_key);
```
Comment (jjyr) [inline gwos-evm/polyjuice-tests/fuzz/gw-syscall-simulator/src/lib.rs]: https://github.com/godwokenrises/godwoken/blob/develop/crates/store/src/state/state_db.rs#L499

The `write_data_hashes` and `read_data_hashes` are only related to these syscalls https://github.com/godwokenrises/godwoken/blob/develop/crates/generator/src/syscalls/mod.rs#L396-L428 we shouldn't insert state key into the `data_hashes`.

Please fix it.

**Review by jjyr — CHANGES_REQUESTED** [review]

File: gwos-evm/polyjuice-tests/fuzz/gw-syscall-simulator/src/lib.rs (hunk 3 above)
Comment (magicalne) [reply gwos-evm/polyjuice-tests/fuzz/gw-syscall-simulator/src/lib.rs]: Fixed. Those data hashes were stored in the wrong place. And run_result cannot be covered in the fuzz testing, so we can just delete them safely.

File: gwos-evm/polyjuice-tests/fuzz/gw-syscall-simulator/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,704 @@
...
+    if let Some(tx) = &HOST.lock().unwrap().tx {
+        store_data(addr, len, 0, tx);
+        SUCCESS
+    } else {
+        GW_ITEM_MISSING
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn gw_load_block_info(addr: *mut c_void, len: *mut u64) -> c_int {
```
Comment (Flouse) [inline gwos-evm/polyjuice-tests/fuzz/gw-syscall-simulator/src/lib.rs:283]: Some extern functions do not require unsafe.

> https://doc.rust-lang.org/book/ch19-01-unsafe-rust.html?highlight=no_mangle#calling-rust-functions-from-other-languages

**Review by Flouse — APPROVED** [review]
LGTM

File: gwos-evm/polyjuice-tests/fuzz/gw-syscall-simulator/src/lib.rs (hunk 1 above)
Comment (magicalne) [reply gwos-evm/polyjuice-tests/fuzz/gw-syscall-simulator/src/lib.rs]: @Flouse Here is the context about turning into `unsafe extern fn`.

File: gwos-evm/polyjuice-tests/fuzz/gw-syscall-simulator/src/lib.rs (hunk 4 above)
Comment (blckngm) [reply gwos-evm/polyjuice-tests/fuzz/gw-syscall-simulator/src/lib.rs:283]: Being `extern` does not require the functions to be `unsafe`, but many of these functions are actually `unsafe`, i.e. it's possible to cause UB if they are called by safe rust functions with incorrect parameters.

**Review by zeroqn — APPROVED** [review]

**Review by jjyr — APPROVED** [review]

**Pull request merged**